src/tools/Repl/Program.cs

```diff
@@ -0,0 +1,601 @@
+            }
+
+            if (enabled.Length == 0)
+            {
+                enabled.Append(" <none>");
+            }
+
+            Console.WriteLine($"Experimental features enabled:{enabled}");
+
+            Console.WriteLine($"Enter Excel formulas.  Use \"Help()\" for details.");
+
+            REPL(Console.In, false);
+        }
+
+        // Pattern match for Set(x,y) so that we can define the variable
+        public static bool TryMatchSet(string expr, out string arg0name, out FormulaValue varValue)
+        {
+            var parserOptions = new ParserOptions { AllowsSideEffects = true };
+
+            var parse = _engine.Parse(expr);
+            if (parse.IsSuccess)
+            {
+                if (parse.Root.Kind == Microsoft.PowerFx.Syntax.NodeKind.Call)
+                {
+                    if (parse.Root is Microsoft.PowerFx.Syntax.CallNode call)
+                    {
+                        if (call.Head.Name.Value == "Set")
+                        {
+                            // Infer type based on arg1. 
+                            var arg0 = call.Args.ChildNodes[0];
+                            if (arg0 is Microsoft.PowerFx.Syntax.FirstNameNode arg0node)
+                            {
+                                arg0name = arg0node.Ident.Name.Value;
+
+                                var arg1 = call.Args.ChildNodes[1];
+                                var arg1expr = arg1.GetCompleteSpan().GetFragment(expr);
+
+                                var check = _engine.Check(arg1expr);
+                                if (check.IsSuccess)
+                                {
+                                    var arg1Type = check.ReturnType;
+
+                                    varValue = check.GetEvaluator().Eval();                                                                        
+                                    _engine.UpdateVariable(arg0name, varValue);
+
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            varValue = null;
+            arg0name = null;
+            return false;
+        }
+
+        public static void REPL(TextReader input, bool echo)
+        {
+            string expr;
+
+            // main loop
+            while ((expr = ReadFormula(input, echo)) != null)
+            {
+                Match match;
+
+                try
+                {
+                    // variable assignment: Set( <ident>, <expr> )
+                    if (TryMatchSet(expr, out var varName, out var varValue))
+                    {
+                        Console.WriteLine(varName + ": " + PrintResult(varValue));
+                    }
+
+                    // formula definition: <ident> = <formula>
+                    else if ((match = Regex.Match(expr, @"^\s*(?<ident>\w+)\s*=(?<formula>.*)$", RegexOptions.Singleline)).Success)
```

Comment:
Would like to get rid of these RegEx, but the corresponding Parse functions are still internal...